File: go/fixture/go-mini/internal/repository/device_repository.go
package repository

import (
	"context"

	"example.com/go-mini/internal/models"
)

// DeviceRepository avoids an import cycle with the services package.
type DeviceRepository interface {
	Get(ctx context.Context, tenant, id string) (models.Device, error)
	Save(ctx context.Context, d models.Device) error
	List(ctx context.Context) ([]models.Device, error)
}

var _ DeviceRepository = (*FileRepo)(nil)

File: py/fixture/py-mini/internal/repository/device_repository.py
"""The repository the services depend on."""

from typing import Protocol

from internal.models.device import Device
from internal.repository.file_repo import FileRepo


class DeviceRepository(Protocol):
    """The store as the services see it; avoids an import cycle with the services package."""

    def get(self, tenant: str, device_id: str) -> Device:
        """Return the device, or raise NotFoundError."""
        ...

    def save(self, d: Device) -> None:
        """Store the device."""
        ...

    def list_devices(self) -> list[Device]:
        """Return every device."""
        ...


_implementers: tuple[type[DeviceRepository], ...] = (FileRepo,)
